### ISSUE TITLE
High RSS usage

### ISSUE THREAD

**Issue by jyotman** [lovell/sharp#714]
Hi, I've seen many similar issues before and I've tried implementing certain measures but have had not luck.

**Scenario** - 
Posting a `.jpg` image of size roughly `300kb` to my Express server. I'm using [multiparty](https://github.com/pillarjs/multiparty) module to receive the image posted from the form as a stream. I pipe this stream to `sharp()` and then to `fs.createWriteStream()`.

**Code** - 
```
const sharp = require('sharp'),
    fs = require('fs'),
    multiparty = require('multiparty');

function saveImage(req, res) {
    const form = new multiparty.Form();

    form.on('part', (part) => {
        if (!part.filename) {
            // console.log('got field named ' + part.name);
            part.resume();
        }

        if (part.filename) {
            console.log(process.memoryUsage());
            part.pipe(sharp()).pipe(fs.createWriteStream('test.jpg'));
            res.end();
        }

        part.on('error', (err) => {
            throw Error(err);
        });
    });

    form.parse(req);
}
```

**Problem** -

`saveImage()` is the function is which receives the POST request. As can be seen in the code, I log the memory usage on every POST call. What I've noticed is that on every POST request my `RSS` value increases by roughly **5-10 mb**. This values keeps on increasing till it reaches a certain threshold around **150 mb** after which it rises very slowly. The problem is that even after reaching such high values, it just remains there. This memory is never released. I left the process running overnight and still the memory consumption after 14 hours was constant at **150 mb**. If I remove the `pipe` to `sharp()` and just write through `fs` like `part.pipe(fs.createWriteStream('test.jpg'));` then this trend disappears and my `RSS` remains stable at around **50 mb** so I'm assuming it definitely something to do with `sharp()`.

Here the output of my logging the `process.memoryUsage()` in each POST request - 

```
{ rss: 52645888,
   heapTotal: 17805312,
   heapUsed: 13611184,
   external: 472783 }
 POST /upload 200 16.620 ms - -
 { rss: 60166144,
   heapTotal: 18853888,
   heapUsed: 14286008,
   external: 1239601 }
 POST /upload 200 3.730 ms - -
 { rss: 65875968,
   heapTotal: 18853888,
   heapUsed: 14478360,
   external: 2005891 }
 POST /upload 200 2.778 ms - -
 { rss: 68939776,
   heapTotal: 18853888,
   heapUsed: 14672288,
   external: 2723029 }
 POST /upload 200 0.814 ms - -
 { rss: 72839168,
   heapTotal: 18853888,
   heapUsed: 14843776,
   external: 3538471 }
 POST /upload 200 1.112 ms - -
 { rss: 76124160,
   heapTotal: 18853888,
   heapUsed: 14966408,
   external: 4255609 }
 POST /upload 200 0.818 ms - -
 { rss: 80297984,
   heapTotal: 18853888,
   heapUsed: 15125680,
   external: 5021899 }
 POST /upload 200 0.845 ms - -
 { rss: 83550208,
   heapTotal: 18853888,
   heapUsed: 15225720,
   external: 5837341 }
 POST /upload 200 1.337 ms - -
 { rss: 88166400,
   heapTotal: 18853888,
   heapUsed: 15416216,
   external: 6603631 }
 POST /upload 200 1.690 ms - -
 { rss: 92119040,
   heapTotal: 18853888,
   heapUsed: 15570144,
   external: 7369921 }
 POST /upload 200 1.512 ms - -
```

Here it can be seen that after just 10 requests my memory usage increased from **50 mb** to **93 mb** for an image size 300 kb. Eventually after more requests it reaches a **150 mb** threshold and then it does not increase much. It kind of becomes constant after that. 

If I use pipelining with sharp and transform a single stream into 3 different `resize()` streams and write them then the problem scales. My `RSS` increases around **25Mb** in each POST request then finally becoming constant at around **300 mb**

I've also tried using `global.gc()` in each POST request but it made no difference at all.

I'm using Node.js version 7.5.0 on Ubuntu 16.04. My system has 12 GB of RAM.

Any help would be greatly appreciated.

**Comment by lovell** [comment]
Hello, the RSS values shown here fit with the pattern of use you describe, especially given they eventually become "constant" based on what I would estimate is peak usage.

Please remember that RSS includes free memory that has not (yet) been returned to the OS. 

If you are writing straight to the filesystem then it'll be more efficient to use `toFile` rather than Streams:
```javascript
  part.pipe(
    sharp().toFile('test.jpg', (err) => { ... })
  );
```
Any Buffer objects created by an instance of `sharp` become eligible for GC when that instance falls out of scope. In your example this would occur only after `req` has also fallen out of scope.

**Comment by jyotman** [comment]
@lovell Thanks for the prompt reply. Actually I'm not writing the files to the disk. This was just a sample I tested to make the example simpler. I'm passing on the image stream to the AWS SDK to upload the file to S3. The stats are almost the same. The code is more like - 

```
const resizeStream = sharp().resize();
part.pipe(resizeStream);
s3.upload({
    Key: 'test.jpg', Body: resizeStream,
    ContentType: 'image/jpeg'
});
```

So I guess I shouldn't be worried about this? Also as you say that any buffer objects created by `sharp` would only be cleaned when `req` has fallen out of scope. So is there any better approach to handle this scenario?

**Comment by lovell** [comment]
Closing as I hope the original question has been answered.
